### PR TITLE
Adds asset file upload

### PR DESCRIPTION
**Pull request by javierarce**
fixes https://github.com/CartoDB/cartodb/issues/11020

This PR enables the upload of assets using direct file upload (of single or multiple files) and also revamps the asset modal window adding tabs.

In addition of those changes, whenever a user uploads a custom asset, the input field of the Fill components shows a generic icon to indicate an image is being used as a marker.

screen shot 2017-02-03 at 14 46 46
screen shot 2017-02-03 at 14 46 58
screen shot 2017-02-03 at 14 47 08
screen shot 2017-02-03 at 14 47 17

### How to do acceptance

0. Enable the `icon-styling` flag.
1. Create a map with some points
2. Style them however you want

- Select an asset from the quick list in the Fill component
- Select an asset from the modal window
- Upload an asset 
- Delete/Select All/Deselect all custom assets from the modal window
- Upload assets (with a URL or using the file browser)

### CODE REVIEW

File: lib/assets/core/javascripts/cartodb3/components/error/error-view.js [hunk 1]
```diff
@@ -21,6 +21,7 @@ module.exports = CoreView.extend({
         desc: _t('components.error.default-desc')
       }
     );
+    this._template = this.options && this.options.template || baseTemplate;
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/error/error-view.js]: `this.options` is always available under a Backbone View, so we don't need the first check.

**Comment by xavijam** [comment]
Hey, have any designer reviewed this @javierarce?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  events: {
+    'click .js-remove': '_onClickRemove',
+    'click .js-select-all': '_onClickSelectAll',
+    'click .js-deselect-all': '_onClickDeselectAll'
+  },
+
+  initialize: function (opts) {
+    checkAndBuildOpts(opts, REQUIRED_OPTS, this);
+
+    this._assets = this.options.assets;
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js]: Is it a collection? If so, I'd call it `_assetsCollection`, in order to be more specific.

Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js]: Or if it is an array, `_assetsArray`.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      assetsCount: this._assets.length,
+      selectedCount: selectedCount,
+      allSelected: selectedCount === this._assets.length
+    }));
+
+    return this;
+  },
+
+  _initBinds: function () {
+    this._assets.bind('add change remove', this.render, this);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js]: Let's try to use `listenTo` if possible (http://stackoverflow.com/questions/16823746/backbone-js-listento-vs-on). If not remember we need to add our `add_related_model` thing.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    this.clearSubViews();
+
+    var selectedCount = this._getSelectedAssetsCount();
+
+    this.$el.html(template({
+      title: this._title,
+      editable: this._editable,
+      assetsCount: this._assets.length,
+      selectedCount: selectedCount,
+      allSelected: selectedCount === this._assets.length
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-header-view.js]: If assets is a collection, `size` method is available.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-item-view.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }));
+
+    this.$el.addClass('AssetsList-item--' + type);
+
+    return this;
+  },
+
+  _onClick: function (e) {
+    this.killEvent(e);
+    this.trigger('selected', this.model);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-item-view.js]: Couldn't we set an attribute called `selected` within the model? With that, we could check the collection changes instead of passing a custom event to the parent.

Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-item-view.js]: Or better, I'm checking that we control that with the `state` attribute. Do we need this event then?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js [hunk 6]
```diff
@@ -2,32 +2,47 @@ var _ = require('underscore');
...
+      opts.size = this.options.size;
+    }
+
+    if (this.options.host !== undefined) {
+      opts.host = this.options.host;
+    }
+
+    if (this.options.ext !== undefined) {
+      opts.ext = this.options.ext;
+    }
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js]: Would it be possible to compact those conditionals with an `_.extend`?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js [hunk 7]
```diff
@@ -38,42 +53,39 @@ module.exports = CoreView.extend({
...
+      url: m.get('public_url'),
+      kind: m.get('kind')
+    });
+
+    this.unselectAssets(m);
   },
 
-  unselectItems: function (m) {
-    this._items.each(function (mdl) {
+  unselectAssets: function (m) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js]: In my opinion, we should add this logic in the collection. Something like:

Collection should listen to any state change from the models, and if that happens, make any stuff it needs (like unselect other items).

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js [hunk 8]
```diff
@@ -38,42 +53,39 @@ module.exports = CoreView.extend({
...
+    this._assets.each(function (mdl) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js]: As I've said `assetsCollection`, `assetsArray`,... (last time I mention it, promise)

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 9]
```diff
@@ -39,80 +64,325 @@ module.exports = CoreView.extend({
     return this;
   },
 
+  _initModels: function () {
+    this._selectedAsset = new cdb.core.Model({
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: We are not using cdb.core.Model anymore (after Builder release). With Backbone.Model is :+1:.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 10]
```diff
@@ -18,19 +32,30 @@ module.exports = CoreView.extend({
...
+
+    this.$el.append(this._assetsTabPaneView.render().el);
+
+    if (this._isLoading()) {
+      this._renderLoading();
+    } else if (this._hasError()) {
+      this._renderError();
+    }
 
     if (this.model.get('image')) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: Why do we disabled add button when image is present in the model? What does it do?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 11]
```diff
@@ -39,80 +64,325 @@ module.exports = CoreView.extend({
...
+      null, {
+        configModel: this._configModel,
+        userModel: this._userModel
+      }
+    );
+  },
+
   _initBinds: function () {
-    this.model.on('change:image', this._onChangeModel, this);
+    this._selectedAsset.on('change:url', this._onChangeSelectedAsset, this);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: As I've said above, if possible, let's try to use `listenTo` (last time I say something about this too, sorry).

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 12]
```diff
@@ -39,80 +64,325 @@ module.exports = CoreView.extend({
...
+    this._stateModel.on('change:status', this.render, this);
+    this._assetsTabPaneView.collection.bind('change', this._onChangeSelectedTab, this);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: If we use this way, we should use `add_related_model` for this collection.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 13]
```diff
@@ -39,80 +64,325 @@ module.exports = CoreView.extend({
...
   },
 
-  _onChangeModel: function () {
-    this.$('.js-add').removeClass('is-disabled');
+  _onClickBack: function (e) {
+    this.killEvent(e);
+    this._stateModel.set('status', '');
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: `this._stateModel.unset('status')` could do the trick?

**Comment by xavijam** [comment]
I can't comment it in the code, but maki icons, simple icons and pin icons disclaimers are not translated, appear in plain english.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 14]
```diff
@@ -39,80 +64,325 @@ module.exports = CoreView.extend({
...
+      this._stateModel.set('status', 'loading');
+    }
+  },
+
+  _onAssetUploaded: function (iconModel) {
+    this._resetFileSelection();
+  },
+
+  _parseResponseText: function (response) {
+    if (response && response.responseText) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: Could you check if `error-parser.js` class module can help you here?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 15]
```diff
@@ -39,80 +64,325 @@ module.exports = CoreView.extend({
...
+    if (!this._selectedAsset.get('url')) {
+      return;
+    }
+
+    this.model.set({
+      image: this._selectedAsset.get('url'),
+      kind: this._selectedAsset.get('kind')
+    });
+
+    this.trigger('change', {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: Why do we need this `change` trigger?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/input-asset-picker-view.js [hunk 16]
```diff
@@ -23,6 +23,18 @@ module.exports = CoreView.extend({
     });
   },
 
+  render: function () {
+    this.clearSubViews();
+    this.$el.html(template());
+    this._initViews();
+    this._initBinds();
+
+    this.$('.js-header').append(this._headerView.render().$el);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/input-asset-picker-view.js]: Better use .el than $el.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/input-asset-picker-view.js [hunk 17]
```diff
@@ -73,12 +73,16 @@ module.exports = CoreView.extend({
     return ramp[this.model.get('index')];
   },
 
-  _onChangeImage: function (url) {
-    this._getRampItem().image = url || '';
+  _onChangeImage: function (data) {
+    var url = data && data.url ? data.url : '';
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/input-asset-picker-view.js]: ` var url = data && data.url || ''`;
` var kind = data && data.kind || ''`;

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/upload-assets-tab.js [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+      this._stateModel.set('status', 'loading');
+    }
+  },
+
+  _onAssetUploaded: function (iconModel) {
+    this._resetFileSelection();
+  },
+
+  _parseResponseText: function (response) {
+    if (response && response.responseText) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/upload-assets-tab.js]: Same function as above, check what I've proposed there 💋 .

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/upload-assets-tab.js [hunk 19]
```diff
@@ -0,0 +1,288 @@
+var _ = require('underscore');
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/upload-assets-tab.js]: Reviewing this "panel", it is the same we used to use in the old upload modal. Are we using the CSS from the old one? If so, we should consider to use a separate one in order to start splitting the CSS from editor and Builder.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+    this.$('.js-assets').append(addAssetButton.render().$el);
+  },
+
+  _initModels: function () {
+    this._stateModel = new Backbone.Model();
+    this.add_related_model(this._stateModel);
+  },
+
+  _initBinds: function () {
+    this._keyDown = this._onKeyDown.bind(this);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: We used to move this selections to the initialize function, as in `context-menu-view.js`, for example.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+
+    this._keyUp = this._onKeyUp.bind(this);
+    $(document).on('keyup', this._keyUp);
+
+    this._stateModel.on('change:status', this.render, this);
+    this._assets.on('change', this._onChangeAssets, this);
+  },
+
+  _getSelectedAsset: function () {
+    var selectedAssets = this._assets.where({ state: 'selected' });
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: findWhere could do the trick here.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+    return selectedAssets && selectedAssets[0];
+  },
+
+  _getSelectedAssetsCount: function () {
+    var selectedAssets = this._assets.where({ state: 'selected' });
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: Same as above.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+    this._shiftKeyPressed = ev.shiftKey;
+  },
+
+  _onKeyUp: function (ev) {
+    this._shiftKeyPressed = ev.shiftKey;
+  },
+
+  _onClickBack: function (e) {
+    this.killEvent(e);
+    this._stateModel.set('status', '');
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: unset?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+  },
+
+  _selectAllAssets: function () {
+    this._assets.each(function (assetModel) {
+      assetModel.set('state', 'selected');
+    });
+  },
+
+  _selectAsset: function (m) {
+    if (this._shiftKeyPressed) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: Ha! nice behaviour for the select and unselect 😗

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+      this._stateModel.set('status', '');
+    } else {
+      this._stateModel.set({
+        error_message: response.statusText,
+        status: 'error'
+      });
+    }
+  },
+
+  _isLoading: function () {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: I've seen several times this function, shouldn't we move it within the proper model? Like a helper one. Same for the below one.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+
+  _removeSelectedAssets: function () {
+    this._stateModel.set('status', 'loading');
+
+    var selectedAssets = this._assets.select(function (asset) {
+      return asset.get('state') === 'selected';
+    }, this);
+
+    _.each(selectedAssets, function (asset) {
+      asset.destroy({
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: I'd not remove any asset until we have a response from the backend. Could we use here `{wait: true}` within the backbone destroy method.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-tab.js [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+        return view;
+      }.bind(this)
+    });
+
+    this.addView(view);
+    this.$el.append(view.render().el);
+  },
+
+  _onFetchAssetsError: function () {
+    console.log('Error');
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-tab.js]: We should render an error message, not only a console.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color-fixed-content-view.js [hunk 28]
```diff
@@ -62,6 +62,7 @@ module.exports = CoreView.extend({
       name: 'file',
       type: this.model.get('image') ? 'image' : 'text',
       label: this.model.get('image') ? this.model.get('image') : _t('form-components.editors.fill.input-color.img'),
+      kind: this.model.get('kind') ? this.model.get('kind') : 'marker',
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color-fixed-content-view.js]: `this.model.get('kind') || 'marker'` ?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js [hunk 29]
```diff
@@ -95,8 +96,16 @@ module.exports = CoreView.extend({
     return false;
   },
 
+  _activatedFlag: function () {
+    return this._userModel && this._userModel.featureEnabled('icon-styling');
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js]: I feel @matallo did something similar or I'm suffering a Dejavù??.

**Comment by javierarce** [comment]
> Hey, has any designer reviewed this @javierarce?

Not in staging. I'll deploy it somewhere so they can see it.

**Comment by javierarce** [comment]
> Hey, has any designer reviewed this @javierarce?

Not in staging, @xavijam. I'll deploy it somewhere so they can see it.

**Comment by xavijam** [comment]
> Not in staging, @xavijam. I'll deploy it somewhere so they can see it.
💋

**Comment by xavijam** [comment]
Ouch, I've just remembered that we have missed one panel within the modal. The organization assets list. Where the user can choose one of the assets uploaded by the organization admin.

**Comment by javierarce** [comment]
I think that was out of the scope of this issue. I don't mind adding it to this ticket since we'll need to implement your suggestions and comments, but in my opinion it would be better to have it as a separate one in order to reduce the size of this PR.

**Comment by xavijam** [comment]
Yep, you are right. Let me create a ticket in order to enable it for organization users.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js (hunk 9 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: Oh, thanks!

**Comment by xavijam** [comment]
@javierarce -> https://github.com/CartoDB/cartodb/issues/11462 🤗 .

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js (hunk 29 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js]: indeed! some lines below https://github.com/CartoDB/cartodb/blob/76822052348fa5ab7029d1f69c63d5a7c25fe64c/lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js#L111

it also checks wether `_imageEnabled` (this comes from `editorAttrs` at editor creation) as only points geometry have image enabled. I'm not sure if this applies here, too, but in any case the `this._userModel && this._userModel.featureEnabled('icon-styling')` part can be put in just one place

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js (hunk 20 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: Oh, I see, to make it more explicit that the view is binding some keys. Nice!

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/upload-assets-tab.js (hunk 19 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/upload-assets-tab.js]: I'm using `.Form-row` and its variations, that are currently being used in the connect dataset form from Builder, and those styles come from `app/assets/stylesheets/common/form-content.css.scss`

File: lib/assets/core/javascripts/cartodb3/components/error/error-view.js (hunk 1 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/error/error-view.js]: I got an undefined error on the specs when I removed that check, I'm not sure why is not defined in this case :|

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js (hunk 6 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-list-view.js]: Yes. That's old code from the original implementation and I think is not needed.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-item-view.js (hunk 5 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/asset-item-view.js]: The reason for this is that changing the state of an asset doesn't only imply the asset itself, but the state of other assets (and this changes if the shift key is pressed). In this case it made more sense to centralice the selection in a common place instead of having it separated in two different locations (view of the item and parent view)

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js (hunk 10 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: We do the opposite, if there's an image selected, we remove the default disabled state of the button.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js (hunk 15 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: To inform the parent (the modal) that an asset selection has been made.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js (hunk 26 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: Yeah, I tried this at first, but I couldn't make it work using wait :/

Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: Ok, talked with you and decided that the first approach was ok for this case (we wait until the deletion is complete and show the list again)

File: lib/assets/core/javascripts/cartodb3/components/error/error-view.js (hunk 1 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/error/error-view.js]: We reviewed this together and found that this.options is not defined.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js (hunk 25 above)
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-list-view.js]: Refactored the upload-tab view to use the upload method from the main modal (assets-view.js)

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js [hunk 30]
```diff
@@ -96,20 +96,16 @@ module.exports = CoreView.extend({
     return false;
   },
 
-  _activatedFlag: function () {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/input-color.js]: isn't it necessary anymore?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-tab.js [hunk 31]
```diff
@@ -63,7 +63,7 @@ module.exports = CoreView.extend({
     this.$el.append(view.render().el);
   },
 
-  _onFetchAssetsError: function () {
-    console.log('Error');
+  _onFetchAssetsError: function (model, response) {
+    this.trigger('error', response);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/user-assets-tab.js]: 🤗

**Comment by javitonino** [comment]
A couple of questions while I am still testing:
- If I style using a custom uploaded icon, the fill component does not show the select icon in the layer view. Intended?
screenshot_20170208_091544
- Should this supports organization assets? They are not shown in the builder view now that it has tabs.

**Comment by xavijam** [comment]
1. Yes, intended, it is ok.
2. Ticket is already created -> https://github.com/CartoDB/cartodb/issues/11462.

**Comment by javitonino** [comment]
I have tested:
- Selecting icons from all pages in the icon modal
- Uploading files from file and URL
- Deleting files
- Select and deselect all

Everything works awesome 👏 

Only a couple design issues (desginer hat on), at least on Firefox:
1. Alignment changes by one pixel when the actions (select all, delete) in the upload tab are shown.
screenshot_20170208_094448
To reproduce, deselect all, stare at your screen, and select one icon in the "Your uploads" tab. You should see the icons move down a pixel.

2. The actions are not consistent with other parts of the interface:
  - The Select/Deselect text is black (should be blue)
  - Both links are not underlined when hovered over
  - No "X items selected" (this probably doesn't make too much sense here)

For reference:
In org assets: 
screenshot_20170208_095024
In builder modal:
screenshot_20170208_095049

They are minor, but I think the first couple design issues are legit.

**Comment by xavijam** [comment]
Great, thanks for discovering that @javitonino. @urbanphes could you check those? Because you checked the feature with @javierarce.

**Comment by javierarce** [comment]
Agree with the style of the link, sorry about that. 

The number of selected items didn't appear in the design, so I'd handle that in a different ticket. And for the 1px issue, I'd open a new ticket.

To speed up the release of this feature, I'll take it and add the required class to fix the link style.

**Comment by javierarce** [comment]
Styling classes added.

**Comment by javierarce** [comment]
@javitonino did you tried enabling/disabling the feature flag?

**Comment by javitonino** [comment]
Yep, there is only a small issue that if you activate it, apply a style with icons, deactivate the FF, the map keeps the icons but you cannot reset them from the style properties. But it is minor (you can reset using cartocss) and only when "downrgrading", so 👍

Everything is hidden when the FF is disabled, and everything works with it enabled ✨

**Comment by javierarce** [comment]
Great, thanks! Promoting it to the next stage.
